File: lua/dapp.lua
local http = require("socket.http")
local ltn12 = require("ltn12")
local json = require("dkjson")

local rollup_server = assert(os.getenv("ROLLUP_HTTP_SERVER_URL"), "missing ROLLUP_HTTP_SERVER_URL")

local function info(...)
    print(string.format(...))
end

local function http_post(url, body)
    local request_body = json.encode(body)
    local response_body = {}
    local result, code = http.request {
        method = "POST",
        url = url,
        source = ltn12.source.string(request_body),
        headers = {
            ["Content-Type"] = "application/json",
            ["Content-Length"] = #request_body
        },
        sink = ltn12.sink.table(response_body)
    }
    if result == nil then error("HTTP POST Request to " .. url .. " failed. " .. code) end
    return code, table.concat(response_body)
end

local handlers = {}
function handlers.advance_state(data)
    info("Received advance request data %s", json.encode(data))
    local lambada_server_addr = os.getenv("LAMBADA_HTTP_SERVER_URL")
    if lambada_server_addr then
        local response, err

        response, err = http.request {
            method = "GET",
            url = lambada_server_addr .. "/open_state",
            headers = { ["Content-Type"] = "application/json" },
        }
        if not response then
            return "Failed to open state: " .. err
        end
        print("State opened successfully.")

        local request_body = "hello world"

        response, err = http.request {
            method = "POST",
            url = lambada_server_addr .. "/set_state/output",
            headers = { ["Content-Type"] = "application/octet-stream" },
            source = ltn12.source.string(request_body),
        }
        if not response then
            return "Failed to set state: " .. err
        end
        print("State set successfully.")

        response, err = http.request {
            method = "GET",
            url = lambada_server_addr .. "/commit_state",
            headers = { ["Content-Type"] = "application/json" },
        }
        if not response then
            return "Failed to commit state: " .. err
        end
        print("State committed successfully.")
    end
    return "accept"
end

function handlers.inspect_state(data)
    info("Received inspect request data %s", json.encode(data))
    -- TODO: add application code here
    return "accept"
end

local mt = {__index = function(t, k) error("Invalid request type: " .. k) end}
setmetatable(handlers, mt)

local finish = {status = "accept"}
while true do
    info("Sending finish")
    local code, response = http_post(rollup_server .. "/finish", finish)
    info("Received finish status %d", code)
    if code == 202 then
        info("No pending rollup request, trying again")
    else
        local rollup_request = json.decode(response)
        local metadata = rollup_request.data.metadata
        finish.status = handlers[rollup_request.request_type](rollup_request.data)
    end
end

File: cpp/dapp.cpp
#include <stdio.h>
#include <iostream>

#include "3rdparty/cpp-httplib/httplib.h"
#include "3rdparty/picojson/picojson.h"

std::string handle_advance(httplib::Client &cli, picojson::value data)
{
    std::cout << "Received advance request data " << data << std::endl;
    const char* lambada_server_url = std::getenv("LAMBADA_HTTP_SERVER_URL");

    if(lambada_server_url) {
        httplib::Client cli(lambada_server_url);
        auto response_open_state = cli.Get("/open_state", {{"Content-Type","application/json"}});
        if (!response_open_state) {
            return "Failed to open state: " + std::to_string(response_open_state ? response_open_state->status : 0);
        }
        std::cout << "State opened successfully." << std::endl;

        auto response_set_state = cli.Post("/set_state/output", "hello world", "application/octet-stream");
        if (!response_set_state) {
            return "Failed to set state: " + std::to_string(response_set_state ? response_set_state->status : 0);
        }
        std::cout << "State set successfully." << std::endl;

        auto response_commit_state = cli.Get("/commit_state", {{"Content-Type","application/json"}});
        if (!response_commit_state) {
            return "Failed to commit state: " + std::to_string(response_commit_state ? response_commit_state->status : 0);
        }
        std::cout << "State committed successfully." << std::endl;
    }
    return "accept";
}

std::string handle_inspect(httplib::Client &cli, picojson::value data)
{
    std::cout << "Received inspect request data " << data << std::endl;
    return "accept";
}

int main(int argc, char **argv)
{
    std::map<std::string, decltype(&handle_advance)> handlers = {
        {std::string("advance_state"), &handle_advance},
        {std::string("inspect_state"), &handle_inspect},
    };
    httplib::Client cli(getenv("ROLLUP_HTTP_SERVER_URL"));
    cli.set_read_timeout(20, 0);
    std::string status("accept");
    std::string rollup_address;
    while (true)
    {
        std::cout << "Sending finish" << std::endl;
        auto finish = std::string("{\"status\":\"") + status + std::string("\"}");
        auto r = cli.Post("/finish", finish, "application/json");
        std::cout << "Received finish status " << r.value().status << std::endl;
        if (r.value().status == 202)
        {
            std::cout << "No pending rollup request, trying again" << std::endl;
        }
        else
        {
            picojson::value rollup_request;
            picojson::parse(rollup_request, r.value().body);
            picojson::value metadata = rollup_request.get("data").get("metadata");
            auto request_type = rollup_request.get("request_type").get<std::string>();
            auto handler = handlers.find(request_type)->second;
            auto data = rollup_request.get("data");
            status = (*handler)(cli, data);
        }
    }
    return 0;
}
